### ISSUE TITLE
AWS Benchmarks sanity check

### ISSUE THREAD

**Issue by wskinner** [horovod/horovod#396]
I know that without RDMA, a performance hit should be expected when training on multiple machines in AWS because gradients must pass through the CPU on their way to the NIC, among other reasons. What I observe is that going from one machine to two is roughly flat in total images per second, and then 90% scaling efficiency resumes when going from 2 to 4 instances. Can anyone else sanity check the numbers I'm seeing?

TF 1.8, CUDA 9.0, ResNet50 from TensorFlow Benchmarks

1 p3.8xlarge:  1399.38 images/sec
1 p3.16xlarge: 2697.29 images/sec
2 p3.16xlarge: 2632.95 images/sec
4 p3.16xlarge: 4772.04 images/sec
8 p3.16xlarge: 9745.38 images/sec

**Comment by richardliaw** [comment]
I used ResNet101, but speedup seems very similar - (p3.16xlarge)

GPUs | 4 | 8 | 16 | 32 | 64 | 128
-- | -- | -- | -- | -- | -- | --
Horovod | 792.41 | 1523 | 1529 | 2958 | 5806 | 11384

**Comment by wskinner** [comment]
@alsrgv 

I ran a few more benchmarks. All were run on ResNet101 with synthetic data. It looks like replacing Kubenet with host networking is good for a 10% speedup, and putting the nodes in a placement group is good for another 13%. Is this about what you expect to see? Interestingly, when the nodes were in the same AZ, but not a placement group, the performance was almost the same as when they were in a placement group together. It's impossible to know, but they probably got scheduled physically close together by chance.

`mpirun -np 16 -x HOROVOD_TIMELINE -x NCCL_MIN_NRINGS=4 -H "172.31.39.57:8,172.31.23.66:8" -mca pml ob1 -mca btl ^openib -mca plm_rsh_args "-p 9999" python scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --batch_size=64 --model=resnet101 --optimizer=momentum --variable_update=horovod --nodistortions --gradient_repacking=8`

| Networking | Images/sec |
| :-------------------------------------------- | ------------ |
| Kubenet, different AZ                                       | 2216.81        |
| Host network, placement group                       | 2739.78       |
| Host network, no placement group, same AZ | 2704.21        |
| Host network, different AZ                               | 2452.37       |

**Comment by alsrgv** [comment]
@wskinner, that looks much more reasonable :-)

**Comment by markmp** [comment]
I've been benchmarking on AWS as well. Specifically using the Version 13 Ubuntu DLAMI which includes horovod in /home/ubuntu/examples/horovod.

I'm training a resnet50 model using the preprocessed and resized imagenet data. (i.e. used both of their scripts). Cmd to mpirun is same in this AWS doc:
https://docs.aws.amazon.com/dlami/latest/devguide/tutorial-horovod.html
E.g:
mpirun.openmpi -np 8 \
--hostfile hostfile --bind-to none --map-by slot -x NCCL_DEBUG=INFO \
-x NCCL_MIN_NRINGS=4 -x LD_LIBRARY_PATH -x PATH  -mca pml ob1 -mca btl ^openib \
python aws_tf_hvd_cnn.py --batch_size=256 --num_epochs=90 --fp16 \
--data_dir train --model resnet50 --log_dir results --display_every 100 --save_interval=3600

2xGPU: ~1481.3 img/sec (only 2 gpus running on a 4 gpu machine - p3.8xlarge
8xGPU: ~5000 img/sec (this is their p3.16xlarge machine, showing about 3.4x speedup)

My main challenge so far is that often when I try mpirun several the of the processes won't start, and it just seems to hang. I've tried changing the hostfile to set the appropriate GPUs, but it seemingly hangs at random..maybe I'll get 3 or 4 that report in , or 5 of 8... And then sometimes it just works. I don't know the cause and it's the main thing I'm trying to figure out.

Also - this was conducted on a gp2 volume that has I think only 160 MB/sec IO capability with 3000 IOPS. I'm worried that I may be disk/io bound which may limit efficiency - and was going to try provisioned IOPS io2 volume to try and see if there is improvement. The TFrecords for imagenet resized are only about 50 GB - so that could be cached in memory on the p3.16x, so not sure if it will help. Curious if others have tested anything with AWS IOP provisioning.

**Comment by Jeffwan** [comment]
> Networking	Images/sec
> Kubenet, different AZ	2216.81
> Host network, placement group	2739.78
> Host network, no placement group, same AZ	2704.21
> Host network, different AZ	2452.37

@wskinner Looks like you submit you job on kubernetes?  How do you setup different environments? Manually or some scripts to build up environments?
